### ISSUE TITLE
onTokenRefresh not gets fired

### ISSUE THREAD

**Issue by peterterbe** [fechanique/cordova-plugin-fcm#216]
Since I have upgraded to v2, I can't get the onTokenRefresh event. On logcat I see the following events:

```
FCMPlugin.js: is created
FCMPlugin Ready OK
Received token refresh
dSMyz8hgbDE:APA91bEPrzyZnFl1xzi3GuaigAGMFeAlOQJQfPSBmv.....

FCMPlugin:{}
FCM Notification callback registered.
```

The FCMPlugin:{} row is a `console.log(JSON.stringify(FCMPlugin))` Shouldn't it contain something? Btw I'm using Ionic framework v2, so I had to do a declare at the beginning of the app to let typescript recognize the FCMPlugin variable. Is it possible that I overwrite the plugin somehow with this declare? But why does it write the last message then? (FCM Notification callback registered.) And it gets a token! Why doesn't onTokenRefresh fired?

Here's my code wich runs at the top of the app:
```
FCMPlugin.onTokenRefresh((t) => {
          console.log("FCM registration token: "+ t);
          //this.user.saveDeviceToken(t);
        });

        FCMPlugin.onNotification((data) => {
          console.log(JSON.stringify(data));
          if(data.wasTapped){
            this.nav.push(ShowmessagePage, {messageid: data['id']});
          } else {
            let confirmAlert = this.alertCtrl.create({
              title: data.title,
              message: data.body,
              buttons: [{
                text: 'Elvet',
                role: 'cancel'
              }, {
                text: 'Megtekint',
                handler: () => {
                  this.nav.push(ShowmessagePage, {messageid: data['id']});
                }
              }]
            });
            confirmAlert.present();
          }
        }, function(msg){
          console.log("FCM Notification callback registered.");
        }, function(err){
          console.log("FCM Notification callback registration error: " + err);
        });
```

**Update:**

If I send a message to the registered device token, it works as expected, so it's just the onTokenRefresh event not get its way into the javascript application.

**Comment by peterterbe** [comment]
Ok, it's working now. It looks like the problem was that the onTokenRefresh event registration in js was after the plugin wanted to call back the event. I have just moved the onTokenRefresh immediately after deviceReady() and now it is working.

**Comment by peterterbe** [comment]
I was just lucky by moving the event subscription forward. It was working once then not again. But this must be a timing issue I think. How can I move FCMPlugin initialization back one second in the row? Will be ugly as hell, but at least I will be sure that the problem is caused by bad timing of deviceReady and FCMPlugin initialization?

**Comment by peterterbe** [comment]
Now it seems to be stable by moving FCM event registration before deviceReady() but there must be some better way, this looks stupid.
Btw, as of ionic v2 rc5 there is a huge 2-5s timeout before deviceReady, so this could be an Ionic framework problem instead.

**Comment by jonathan-chin** [comment]
hey @peterterbe do you mind sharing your code? I'm also working with ionic v2 and am trying to get this plugin working.

I added it with google-services.json and it seemed to work. but here's one of my pages:

```
import {Component} from "@angular/core";
import {NavController} from "ionic-angular";
import {MapPage} from "../map/map";
import {SharePage} from "../share/share";

var FCMPlugin;

@Component({
    selector: "page-home",
    templateUrl: "home.html"
})


export class HomePage{
    private pages: Object = {
        map: {page: MapPage, navOptions: {}},
        event: {page: SharePage, navOptions: {type: "event"}},
        person: {page: SharePage, navOptions: {type: "person"}},
    }

    constructor(private nav: NavController){
        console.log(FCMPlugin);

    }

    goto(page){
        this.nav.push(this.pages[page].page, this.pages[page].navOptions).catch((error) => {
            // up to rejecting class to provide error feedback                                                                                                                                
        });
    }
}
```

console.log(FCMPlugin) returns undefined when run on my phone (I remote debug it through Chrome).
that's not expected, right?

**Comment by peterterbe** [comment]
@jonathan-chin I had to add `declare var FCMPlugin;` at the beginning of the app.component.js
Then I have an initializePush() method that registers my callback methods as described in this project's readme.

My problem was that the this.platform.ready() was happened too late, so when my application registered its onTokenRefresh callback it was too late. (I suppose.) I don't know how stable is my solution but I have moved the initializePush() method with FCM callback registrations outside platformReady(), right in my constructor:
```
constructor(public platform: Platform, private user:UserProvider,
            private alertCtrl: AlertController, private appCfg:AppConfig,
            private event: Events) {
    this.initializePush();

    this.initializeApp();
```

FCMPlugin should be an object. For me it was undefined only when debugging in chrome with `ionic serve`, so it looks like your debugging method messes up something.

**Comment by jonathan-chin** [comment]
@peterterbe I figured it out. I'm developing as a progressive app, on web and device. I forgot to put back in cordova.js when pushing to a phone. once I did, I had no trouble with the plugin.

silly mistake on my part!

**Comment by elewin** [comment]
I'm running into the same issue too, but on Ionic v1. I'm putting FCMPlugin.onTokenRefresh() in the callback of $ionicPlatform.ready(), and while onNotification works if I manually copy+paste the token from the console, I can't get the app to see the token itself on its own

**Comment by elewin** [comment]
Ok, I think I have figured out what many people are running into:
If they call onTokenRefresh too late, then the token event fires before the callback is registered so nothing happens. If they call getToken too early, they get a null token because a token has not been issued yet. The key is to use both methods: onTokenRefresh to catch late tokes, and getToken to catch early/pre-existing tokens.

**Comment by peterterbe** [comment]
@elewin Yes, me too finally had to to use the getToken() method at the end of push initialization in case of the onTokenRefresh callback was fired too early (event registration is at the beginning).

**Comment by larssn** [comment]
Ugh, this seems a hacky way to fix this problem :(

**Comment by larssn** [comment]
Looking at https://github.com/fechanique/cordova-plugin-fcm/blob/master/www/FCMPlugin.js#L22 I can't really see how onTokenRefresh would ever work, as it (and its sister function), doesn't seem to interact with Cordova at all.

Can you guys confirm that I'm not crazy. It doesn't invoke the `cordova/exec` function.

**Comment by larssn** [comment]
I assume that onTokenRefresh can be missed, if the app is minimized when the native code fires. If so, that would require the background mode plugin; which in turn means vastly less battery life.
